### ISSUE TITLE
Add support for Raspberry Pico W(ifi)

### ISSUE THREAD

**Issue by Brandl** [adafruit/circuitpython#6558]
While the pico is listed on the Downloads page, it is not entirely clear how to use the newly added wifi connectivity with circuitpython.

According to adafruit:
https://www.adafruit.com/product/5526

    At the time of launch, only MicroPython has WiFi support

Being able to drop in a new pico W into existing projects to add wireless connectivity would be awesome. Looking forward to updates on this!

**Comment by richsad** [comment]
I do not currently do my own builds of CP but and create core modules but will begin doing that. Won't take long. I eat technology for breakfast ;-)

**Comment by ratarchivos** [comment]
how can i create a ap(local host) with pico w?

**Comment by tannewt** [comment]
> how can i create a ap(local host) with pico w?

You can't currently from CircuitPython.

@richsad Please let us know when you've started working on this. I want to make sure folks don't duplicate work.

**Comment by dhalbert** [comment]
@richsad - there is a MicroPytyhon Pico W build, including wifi: https://micropython.org/download/rp2-pico-w/

As mentioned above, the relevant commit in MicroPython is https://github.com/micropython/micropython/commit/50e46552c0c83b0c8570aba098a0ba4e68ae9eac

The library that supports the module is here: https://github.com/georgerobotics/cyw43-driver.git

**Comment by richsad** [comment]
Amended: I see Dan has provided the link I needed. Awesome. 
I have two Pico-W (aka rpi2-PICO-W) boards in house and have a build system set aside on an Rpi4b 4GB or if needed an Intel linux machine. Thank you for the info @dhalbert . Maybe the fact I was searching at 3am may be partly why I couldn't find it!

**Comment by dhalbert** [comment]
I think you will vastly prefer the Intel Linux machine due to build speed. Don't forget `make -j<n>`: https://learn.adafruit.com/building-circuitpython/build-circuitpython#use-all-your-cpus-when-building-3086075

**Comment by nbendall34** [comment]
@richsad glad Dan was able to provide you with the info needed and thank you for working on this :)

Do we have a timeframe for this release?

**Comment by ckocyigit** [comment]
Hello there,

are there any updates or is there a approximate timeframe for the change?

Thank you!

**Comment by ckocyigit** [comment]
Hello there,

are there any updates or is there a approximate timeframe for the change?

Thank you!

**Comment by Myaanee** [comment]
I hate to be that person but is there any Update on this @richsad ? I'm sure there's tons of people looking forward to wifi support on this board

**Comment by CWinicki** [comment]
Just emailed @richsad and he replied that has not made much progress and is not currently able to continue working on it. Just wanted to send along the message since hey may not get emails from this thread.

**Comment by dhalbert** [comment]
Adafruit is working on this: see an early progress report here: https://www.youtube.com/watch?v=-WbU1tmKuCw.

**Comment by sgbaird** [comment]
Would simplify a lot to be able to use circuit python, such as using the adafruit io package instead of low level mqtt

**Comment by gabriel-wainmann** [comment]
Hi

Would love an update on the WiFi capabilities of the PicoW with CircuitPython.
Any ballpark on when this will see the light of day?

Thanks!

**Comment by Neradoc** [comment]
Some update: It connects. See the latest Python on hardware weekly video:

[![Python on hardware weekly](https://img.youtube.com/vi/9u_CYthLWbg/0.jpg)](https://youtu.be/9u_CYthLWbg?t=164)

**Comment by BjornTheProgrammer** [comment]
I cannot seem to find the source code for the development of this. Where could I find it? I'm highly interested in the implementation. Would greatly appreciate it if I could view it!

**Comment by anecdata** [comment]
Would be really nice to have access to UF2 artifacts ;-)

**Comment by DeadSix27** [comment]
> Would be really nice to have access to UF2 artifacts ;-)

I think they're developing it private, but I agree, I'd love to have some pre-release builds to tinker with

**Comment by jepler** [comment]
Hiya!

My work in progress, which is what @ladyada has shown in a couple of videos, is in the open at [https://github.com/jepler/circuitpython/tree/:pie::cow:](https://github.com/jepler/circuitpython/tree/%F0%9F%A5%A7%F0%9F%90%AE)

It really doesn't do anything _useful_ yet, we haven't identified any ways to split the work with other folks, and it's far from ready to use or give feedback on, so publicizing where the code lives or providing UF2s isn't a priority. I'm glad to see the videos have built up anticipation, though!

**Comment by jepler** [comment]
Once build artifacts are available on #6933 I welcome feedback about what is/is not working from others. Please keep in mind the caveats in the initial comment of that PR.

**Comment by DeadSix27** [comment]
`Quick question, is ssl on the todo list? As you said, its rather important.

**Comment by jepler** [comment]
Yup, agreed that ssl is important.

**Comment by DeadSix27** [comment]
> Yup, agreed that ssl is important.

also thanks for all your work! I'll be testing as soon the PR build worker has finished!

**Comment by bill88t** [comment]
Built latest commit on my ovahclocked af Pi400, after tagging it `8.0.0-picow-01-887c259`

```
Adafruit CircuitPython 8.0.0-picow-01-887c259 on 2022-09-22; Raspberry Pi Pico W with rp2040
```

The scans always find some networks duplicated:
```
Pi400
Pi400
Pi400
Pi400
Forthnet-AKzTk
Pi400
Feline34
WIND_2.4G_422DEA
COSMOTE-403528
WIND_C02F18
WIND_C02F18
WIND_C02F18
WIND_2.4G_9DE02B
WIND_2.4G_9DE02B
WIND_2.4G_9DE02B
WIND_C02F18
```

Created a [code.py](https://git.port1.cf/bill88t/Picow-testing/src/branch/master/code_picow.py) to test everything.

After about 20 executions got a `OverflowError: overflow converting long int to machine word` when the RTC() was being set.

Seems like an one off, since on next reload it worked just fine.
But what happens once is bound to happen again.

After around 150 runs it just hanged on the wifi scanning.
And now regardless of if I powercycle it, reload it, reset it. It always just does two-three runs and then hangs on the wifi scanning.
The hang occurs on the `for` line.

**Comment by jepler** [comment]
> The scans always find some networks duplicated:

I agree, though I saw the same thing with the pico-examples network scan program. If this is a concern for you, you can report it upstream.

**Comment by DeadSix27** [comment]
Is there a reason rainbowio is excluded? Just curious, since afaik its just a basic function?

**Comment by bill88t** [comment]
Hm, my manual build has it. Perhaps he disabled it for rom storage?

**Comment by jepler** [comment]
> Is there a reason rainbowio is excluded? Just curious, since afaik its just a basic function?

It's disabled for an obscure samd21 port, not for pico w.

**Comment by sgbaird** [comment]
Is there a simple example/set of instructions for connecting to WiFi via Pico W in the `8.0.0` beta release?

**Comment by anecdata** [comment]
@sgbaird 
wifi works basically the same as espressif (esp32-s2), though not all features or properties are present at present
```py
import wifi
import socketpool
import adafruit_requests
from secrets import secrets

TEXT_URL = "http://wifitest.adafruit.com/testwifi/index.html"

pool = socketpool.SocketPool(wifi.radio)
requests = adafruit_requests.Session(pool, None)

wifi.radio.connect(secrets["ssid"], secrets["password"])

with requests.get(TEXT_URL) as r:
    print(f"{r.status_code} {r.reason.decode()} {r.content}")
```
If you get an `ImportError` or `NotImplementedError`, it's probably not there yet. Other exceptions (e.g., `OSError`, `ConnectionError`): `try` / `except` and retry.

**Comment by sgbaird** [comment]
@anecdata thanks for the rapid reply! Figured out that the Pico W firmware isn't installable via Thonny yet (4.0.1), but installing directly from CircuitPython works: https://circuitpython.org/board/raspberry_pi_pico_w/

**Comment by jepler** [comment]
Thanks @anecdata -- that's spot on!

The new work on pico w implements these modules: `wifi` and `socketpool`, which are intended to be compatible with our Espressif port except for the NotImplemented items. This, in turn, allows modules like `adafruit_requests` to work unaltered. So, basically, try adapting examples for esp32-s2 and the rest of the Espressif family.

We're waiting on the outcome of trying to add ssl before writing any guides about this on adafruit.com.

**Comment by jepler** [comment]
(It _does_ implement a new module, [cyw43](https://docs.circuitpython.org/en/latest/shared-bindings/cyw43/index.html), but in most cases it's not necessary to import or use this module)

**Comment by sgbaird** [comment]
Having some trouble with MQTT and Pico W https://github.com/adafruit/circuitpython/issues/6988. Used `while True: try: ... except: ...` retries also.

**Comment by DeadSix27** [comment]
Not sure if this is related to the implementation, but whenever I use `adafruit_httpserver` and run a basic example, like reading a i2c sensor on demand (via http request) all my browser receives is a "ERR_CONNECTION_RESET"

Now the same setup but using `ampule` as server works just fine, is there something I'm missing?


On a side note, I also sometimes encounter `ConnectionError: Unknown failure 205` during radio connect, but I cant rule out my network setup being the cause.
